### PR TITLE
Add optimization pass to remove bp stores and restores

### PR DESCRIPTION
**Pull request by lockshaw**

### CODE REVIEW

**Comment by ltfish** [comment]
Good job!

Linting failed :) (https://travis-ci.org/angr/angr/jobs/498982938) Fix linting and I'll click merge!

**Comment by ltfish** [comment]
Is it still WIP? It seems pretty complete to me...

**Comment by lockshaw** [comment]
> 
> 
> Is it still WIP? It seems pretty complete to me...

I need to add tests

**Comment by ltfish** [comment]
Ping? Any progress?

**Comment by lockshaw** [comment]
Just added a basic test case and updated it a bit (improving stack pointer tracking made a previously non-functioning part of the decompiler work, causing the bp save to look different). Should be ready to merge in now.

File: tests/test_bp_save_optimization.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import ailment
+
+test_location = str(
+    os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), '../../binaries/tests'))
+
+
+def test_bp_save_fauxware(arch):
+    p = angr.Project(
+        test_location + '/{}/fauxware'.format(arch), auto_load_libs=False)
```
Comment (ltfish) [inline tests/test_bp_save_optimization.py]: Nitpick: Can you use `os.path.join()` instead?

Comment (lockshaw) [reply tests/test_bp_save_optimization.py]: Yup!

**Comment by ltfish** [comment]
LGTM. Thanks. Will merge once tests pass.

**Comment by ltfish** [comment]
Thanks!

**Pull request merged**